src/plugins/tlsmon/tlsmon.cpp

```diff
@@ -0,0 +1,293 @@
+    }
+
+    // Find the buffer containing client random.
+    auto it = std::find_if(ncrypt_buffers.begin(), ncrypt_buffers.end(), [&](const auto& e)
+    {
+        return e.buffer_type == tlsmon_priv::NCRYPTBUFFER_SSL_CLIENT_RANDOM;
+    });
+    if (it == ncrypt_buffers.end())
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    // And finally read it.
+    std::array<char, tlsmon_priv::CLIENT_RANDOM_SZ> client_random = std::array<char, tlsmon_priv::CLIENT_RANDOM_SZ>();
+    ctx.addr = (addr_t) it->buffer;
+    if (VMI_SUCCESS != vmi_read(lg.vmi, &ctx, client_random.size(), client_random.data(), nullptr))
+    {
+        plugin->destroy_trap(info->trap);
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+
+    // Output retrieved data in hex format.
+    std::string master_key_str = tlsmon_priv::byte2str((unsigned char*)master_secret.master_key, tlsmon_priv::MASTER_KEY_SZ);
+    std::string client_random_str = tlsmon_priv::byte2str((unsigned char*)client_random.data(), tlsmon_priv::CLIENT_RANDOM_SZ);
+    fmt::print(plugin->m_output_format, "tlsmon", drakvuf, info,
+               keyval("client_random", fmt::Qstr(client_random_str)),
+               keyval("master_key", fmt::Qstr(master_key_str))
+              );
+
+    plugin->destroy_trap(info->trap);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+
+/**
+ * Sets a trap on return from SslGenerateMasterKey function to obtain the
+ * calculated master key.
+ */
+static
+event_response_t ssl_generate_master_key_cb(drakvuf_t drakvuf, drakvuf_trap_info* info)
+{
+    tlsmon* plugin = static_cast<tlsmon*>(info->trap->data);
+
+    auto trap = plugin->register_trap<ssl_generate_master_key_result_t>(
+                    info,
+                    ssl_generate_master_key_ret_cb,
+                    breakpoint_by_dtb_searcher(),
+                    "SslGenerateMasterKey"
+                );
+    if (!trap)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    auto params = get_trap_params<ssl_generate_master_key_result_t>(trap);
+    params->set_result_call_params(info, drakvuf_get_current_thread(drakvuf, info));
+    params->master_key_handle_addr = drakvuf_get_function_argument(drakvuf, info, 4);
+    params->parameter_list_addr = drakvuf_get_function_argument(drakvuf, info, 7);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+
+/**
+ * Sets a hook on running lsass process. In Windows, processes that want to
+ * establish TLS connection with Schannel API, do so by using lsass under the
+ * hood. This way, lsass will perform TLS handshake on behalf of the process
+ * initiating the connection and secrets will never leave lsass's memory.
+ */
+void tlsmon::hook_lsass(drakvuf_t drakvuf)
+{
+    addr_t lsass_base = 0;
+    if (!drakvuf_find_process(drakvuf, ~0, "lsass.exe", &lsass_base))
+        return;
+    drakvuf_request_userhook_on_running_process(drakvuf, lsass_base, "ncrypt.dll", "SslGenerateMasterKey", ssl_generate_master_key_cb, this);
```

Comment:
What about:
https://github.com/tklengyel/drakvuf/blob/b339f691bec4ccba583ff5d910f4672029484bff/src/libusermode/running.cpp#L495

I belive this implicitly requires that we have some other plugin like apimon started before tlsmon, in order to initialize `userhook` instance. The tlsmon alone (`-a tlsmon`) would crash because `userhook` singleton was not initialized.

Reply:
nevermind, I see this problem is addressed within this PR